File: vco_lib/gateway_usage.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2026 VibeCoded Tools
"""Read the model gateway's subscription usage windows — the launcher's bridge.

The gateway computes and caches the windows (``model_router.usage_windows``,
served on ``GET /usage/windows``). This module is the one client the launcher
spawns for them (``python -m vco_lib.gateway_usage``), for the reason every
gateway call from the launcher goes through Python: the host token must never
cross the Rust process (``launcher/src-tauri/src/commands/model_gateway.rs``,
"The host token never crosses this process"). The token is read here from the
gateway's own file, put in one request header, and never printed.

Port and token resolution are NOT re-implemented: both come from
:mod:`vco_lib.vscode_settings` (``resolve_gateway_ports`` /
``resolve_host_token``), the same answer the panel wiring uses.

stdout is a machine contract: exactly one JSON object —
``{"ok": true, "snapshot": {...}}`` or ``{"ok": false, "reason": <word>,
"message": <text>}`` — and the exit code is 0 either way, because "the
gateway is not running" is an answer the card renders, not a crash.
"""

from __future__ import annotations

import argparse
import json
import sys
import urllib.error
import urllib.request
from typing import Any, Optional, Sequence

#: The gateway route. MUST MATCH ``create_app`` in
#: ``claude_mcp_servers/model_router/server.py``.
ROUTE = "/usage/windows"

#: Seconds. Loopback, answered from the gateway's memory — generous only so a
#: loaded machine does not read as "unreachable".
DEFAULT_TIMEOUT_S = 3.0

REASON_NO_TOKEN = "no_token"
REASON_UNREACHABLE = "unreachable"
REASON_UNAUTHORISED = "unauthorised"
REASON_BAD_ANSWER = "bad_answer"
REASON_OUTDATED = "outdated_gateway"


def _failure(reason: str, message: str) -> dict:
    return {"ok": False, "reason": reason, "message": message}


def fetch_windows(*, timeout_s: float = DEFAULT_TIMEOUT_S) -> dict:
    """Ask the running gateway for its usage snapshot. Never raises."""
    # Imported here so `--help` works on a machine whose gateway package is
    # not importable, and so the failure is reported as data.
    from vco_lib.vscode_settings import (
        SettingsRefused,
        resolve_gateway_ports,
        resolve_host_token,
    )

    try:
        token = resolve_host_token()
    except (SettingsRefused, ImportError) as exc:
        return _failure(REASON_NO_TOKEN, str(exc))
    if not token:
        return _failure(REASON_NO_TOKEN, "the gateway's host token file is empty")
    port = resolve_gateway_ports()[0]
    request = urllib.request.Request(
        f"http://127.0.0.1:{port}{ROUTE}",
        headers={"Authorization": f"Bearer {token}"},
    )
    # A loopback call must not be routed through a user's HTTP proxy.
    opener = urllib.request.build_opener(urllib.request.ProxyHandler({}))
    try:
        with opener.open(request, timeout=timeout_s) as resp:  # noqa: S310 — loopback only
            payload: Any = json.loads(resp.read())
    except urllib.error.HTTPError as exc:
        if exc.code == 401:
            return _failure(
                REASON_UNAUTHORISED,
                f"the gateway on port {port} refused the host token (HTTP 401)",
            )
        if exc.code == 404:
            # The route arrived in 0.2.97; a gateway started before the
            # update is still serving the previous release's routes.
            return _failure(
                REASON_OUTDATED,
                f"the gateway on port {port} has no {ROUTE} route — it is "
                "running an older version; restart it to load the updated one",
            )
        return _failure(REASON_BAD_ANSWER, f"the gateway on port {port} answered HTTP {exc.code}")
    except (urllib.error.URLError, OSError) as exc:
        return _failure(REASON_UNREACHABLE, f"no gateway answered on port {port}: {exc}")
    except ValueError:
        return _failure(REASON_BAD_ANSWER, f"the gateway on port {port} did not answer JSON")
    if not isinstance(payload, dict) or not isinstance(payload.get("vendors"), list):
        return _failure(
            REASON_BAD_ANSWER,
            f"the service on port {port} answered without a usage snapshot "
            "(an older gateway, or not a gateway)",
        )
    return {"ok": True, "port": port, "snapshot": payload}


def _build_arg_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="python -m vco_lib.gateway_usage",
        description="Print the model gateway's subscription usage windows as JSON.",
    )
    parser.add_argument(
        "--timeout", type=float, default=DEFAULT_TIMEOUT_S,
        help=f"seconds to wait for the gateway (default {DEFAULT_TIMEOUT_S})",
    )
    return parser


def main(argv: Optional[Sequence[str]] = None) -> int:
    args = _build_arg_parser().parse_args(argv)
    sys.stdout.write(json.dumps(fetch_windows(timeout_s=args.timeout)) + "\n")
    return 0


if __name__ == "__main__":  # pragma: no cover - exercised via subprocess
    raise SystemExit(main())

File: launcher/src-tauri/src/commands/gateway_usage.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 VibeCoded Tools
//! Subscription usage windows for the home-page card — the launcher half.
//!
//! The gateway computes and caches the windows (`model_router.usage_windows`,
//! `GET /usage/windows`); `python -m vco_lib.gateway_usage` reads them with
//! the host token. This command only spawns that bridge and relays its one
//! JSON object, for the reason every gateway call from the launcher goes
//! through Python (A>B>C, rule A): the host token never crosses this process
//! (see `model_gateway.rs`, "The host token never crosses this process").
//!
//! The card polls, so the not-running case is answered HERE from the pid
//! file without spawning an interpreter — the same short-circuit
//! `gateway_freshness` takes.

use tauri::command;

use crate::commands::model_gateway::{
    gateway_usage_command, probe_process, python_or_err, run_to_completion_within, ProcessState,
};

/// The bridge makes one loopback GET answered from the gateway's memory
/// (its own timeout is 3 s); the rest is interpreter start-up.
const BRIDGE_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(20);

/// The answer for a gateway that is not running. Same shape as the bridge's
/// own failures (`vco_lib.gateway_usage._failure`), so the card has one
/// branch for "no numbers".
pub fn not_running_payload() -> serde_json::Value {
    serde_json::json!({
        "ok": false,
        "reason": "not_running",
        "message": "the model gateway is not running",
    })
}

/// Parse the bridge's stdout. Its stdout is a machine contract — one JSON
/// object, exit 0 even for "unreachable" — so anything else is a broken
/// install and is reported as an error carrying the bridge's stderr.
pub fn parse_bridge_output(
    code: i32,
    stdout: &str,
    stderr: &str,
) -> Result<serde_json::Value, String> {
    match serde_json::from_str::<serde_json::Value>(stdout.trim()) {
        Ok(v) if v.get("ok").and_then(|ok| ok.as_bool()).is_some() => Ok(v),
        Ok(_) => Err(format!(
            "vco_lib.gateway_usage exited {} with JSON that is not a usage answer: {}",
            code,
            stderr.trim()
        )),
        Err(e) => Err(format!(
            "vco_lib.gateway_usage exited {} and did not return JSON ({}): {}",
            code,
            e,
            stderr.trim()
        )),
    }
}

fn fetch_blocking() -> Result<serde_json::Value, String> {
    if probe_process().0 != ProcessState::Running {
        return Ok(not_running_payload());
    }
    let python = python_or_err()?;
    let root = crate::commands::installer::find_local_repo_root().ok();
    let cmd = gateway_usage_command(&python, root.as_deref());
    let (code, stdout, stderr) =
        run_to_completion_within(cmd, "vco_lib.gateway_usage", BRIDGE_TIMEOUT)?;
    parse_bridge_output(code, &stdout, &stderr)
}

/// `{"ok": true, "port": N, "snapshot": {...}}` or
/// `{"ok": false, "reason": ..., "message": ...}`.
#[command]
pub async fn model_gateway_usage_windows() -> Result<serde_json::Value, String> {
    tauri::async_runtime::spawn_blocking(fetch_blocking)
        .await
        .map_err(|e| format!("usage-windows task failed: {}", e))?
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_bridge_success_and_failure_shapes_both_parse() {
        let ok = parse_bridge_output(
            0,
            r#"{"ok": true, "port": 11460, "snapshot": {"vendors": []}}"#,
            "",
        )
        .unwrap();
        assert_eq!(ok["port"], 11460);
        let failed = parse_bridge_output(
            0,
            r#"{"ok": false, "reason": "unreachable", "message": "m"}"#,
            "",
        )
        .unwrap();
        assert_eq!(failed["reason"], "unreachable");
    }

    /// F4: a broken install exits 1 WITH its JSON answer. It must reach the
    /// card as that answer (so the card can name the problem), not be
    /// discarded for the non-zero exit.
    #[test]
    fn a_broken_install_answer_survives_its_nonzero_exit() {
        let v = parse_bridge_output(
            1,
            r#"{"ok": false, "reason": "broken_install", "message": "re-run install.py"}"#,
            "vco_lib.gateway_usage: broken install",
        )
        .unwrap();
        assert_eq!(v["reason"], "broken_install");
    }

    #[test]
    fn anything_else_is_an_error_carrying_stderr() {
        let err = parse_bridge_output(
            1,
            "Traceback (most recent call last)",
            "ImportError: vco_lib",
        )
        .unwrap_err();
        assert!(err.contains("ImportError: vco_lib"), "{err}");
        let err = parse_bridge_output(0, r#"{"vendors": []}"#, "").unwrap_err();
        assert!(err.contains("not a usage answer"), "{err}");
    }

    #[test]
    fn not_running_has_the_bridge_failure_shape() {
        let v = not_running_payload();
        assert_eq!(v["ok"], false);
        assert_eq!(v["reason"], "not_running");
        assert!(v["message"].as_str().is_some_and(|m| !m.is_empty()));
    }
}
